src/components/files/archive.ts

```diff
@@ -0,0 +1,137 @@
+/*
+Script to archive files older than a certain threshold
+Please see https://github.com/openmsupply/conforma-server/discussions/1059 for details
+*/
+
+import DBConnect from '../databaseConnect'
+import { DateTime } from 'luxon'
+import { move, mkdirp, readJSON, writeJSON } from 'fs-extra'
+import path from 'path'
+import { clearEmptyDirectories } from '../utilityFunctions'
+import { ARCHIVE_FOLDER, ARCHIVE_SUBFOLDER_NAME, FILES_FOLDER } from '../../constants'
+import config from '../../config'
+import { nanoid } from 'nanoid'
+
+const isManualArchive: Boolean = process.argv[2] === '--archive'
+const param = Number.parseInt(process.argv[3])
+const archiveDays = Number.isNaN(param) ? undefined : param
+
+export interface ArchiveInfo {
+  timestamp: number
+  archiveFolder: string
+  uid: string
+  prevArchiveFolder: string | null
+  prevUid: string | null
+  numFiles: number
+}
+
+export interface ArchiveData {
+  archives: { [key: string]: ArchiveInfo }
+  history: ArchiveInfo[]
+}
+
+export const archiveFiles = async (days: number = config.archiveFileAgeMinimum ?? 7) => {
+  console.log(
+    DateTime.now().toLocaleString(DateTime.DATETIME_SHORT_WITH_SECONDS),
+    `Archiving files older than ${days} days...`
+  )
+
+  // Load archive history
+  let archiveData: ArchiveData
+  try {
+    archiveData = await readJSON(path.join(ARCHIVE_FOLDER, 'archive.json'))
+  } catch {
+    archiveData = { archives: {}, history: [] }
+  }
+
+  const { archives, history } = archiveData
+
+  // TO-DO: Verify integrity of existing archive
+
+  // Get file list
+  const files = await DBConnect.getFilesToArchive(days)
+  if (files.length === 0) {
+    console.log('Nothing to archive')
+    return null
+  }
+  // Create archive subfolder
+  const timestamp = DateTime.now()
+  const timestampString = timestamp.toFormat('yyyy-LL-dd_HH-mm-ss')
+  const uid = nanoid()
+  const folderName = `${timestampString}_${uid.slice(0, 6)}`
+  const archivePath = path.join(ARCHIVE_SUBFOLDER_NAME, folderName)
+  await mkdirp(path.join(FILES_FOLDER, archivePath))
+
+  // Move files
+  for (const file of files) {
+    const newFilePath = path.join(archivePath, 'files', file.file_path)
+    try {
+      await move(path.join(FILES_FOLDER, file.file_path), path.join(FILES_FOLDER, newFilePath))
+      file.archive_path = path.join(archivePath, 'files')
+    } catch {
+      console.log('Problem moving', file.file_path)
+    }
+    const shouldMoveThumbnail = !file.thumbnail_path.startsWith(config.genericThumbnailsFolderName)
```

Comment:
We don't want to move thumbnails that use the generic system thumbnails (e.g. PDFs)